src/Enzyme.jl

```diff
@@ -1006,64 +1006,75 @@ end
     return (one(x),)
 end
 
-"""
-    gradient(::ReverseMode, f, x)
+@inline function chunkedonehot(x, ::Val{chunk}) where chunk
+    sz = length(x)
+    num = ((sz + chunk - 1) ÷ chunk)
+    ntuple(Val(num)) do i
+        Base.@_inline_meta
+        onehot(x, (i-1)*chunk+1, i == num ? sz : (i*chunk) )
+    end
+end
 
-Compute the gradient of a real-valued function `f` using reverse mode.
-This will allocate and return new array `make_zero(x)` with the gradient result.
+@inline function chunkedonehot(x::AbstractFloat, ::Val{chunk}) where chunk
+    return ((one(x),),)
+end
 
-Besides arrays, for struct `x` it returns another instance of the same type,
-whose fields contain the components of the gradient.
-In the result, `grad.a` contains `∂f/∂x.a` for any differential `x.a`,
-while `grad.c == x.c` for other types.
+@inline tupleconcat(x) = x
+@inline tupleconcat(x, y) = (x..., y...)
+@inline tupleconcat(x, y, z...) = (x..., tupleconcat(y, z...)...)
 
-Examples:
+@inline function derivative(mode::ForwardMode, f, x; shadow=onehot(x))
+    values(only(autodiff(mode, f, BatchDuplicatedNoNeed, BatchDuplicated(x, shadow))))
+end
+@inline function derivative_deferred(mode::ForwardMode, f, x; shadow=onehot(x))
+    values(only(autodiff_deferred(mode, f, BatchDuplicatedNoNeed, BatchDuplicated(x, shadow))))
+end
 
-```jldoctest gradient
-f(x) = x[1]*x[2]
+@inline _chunkcheck(::Val{0}) = throw(ArgumentError("Cannot differentiate with a batch size of 0"))
+@inline _chunkcheck(::Val) = nothing
 
-grad = gradient(Reverse, f, [2.0, 3.0])
+@inline function derivative(mode::ForwardMode, f::F, x::X, ::Val{chunk};
```

Comment:
what is the semantics for derivative? I would've thought it be one input one output. If it needs multiple inputs though, why not just have this call gradient?

Reply:
Ah I see, those now call this

Reply:
Yeah, the basic idea is that all `gradient` and `jacobian` do is re-arrange output of this, so it should be its own function.  The name `derivative` is perhaps too grandiose, I didn't have any more sophisticated idea of what it should be than the thing the returns of which are manipulated by those.

Reply:
Okay if that's the case, it would definitely be helpful to have a docstring stating what the semantics are (as distinct from others, since I can see people trying to use derivative and then getting confused).

We could also pick a different name here, like unstructured_derivative or something (which would likely dissuade folks from using it in contrast to gradient or jacobian).

Reply:
Also this function will fail on a chunk size of 1 (duplicated, not batchduplicated should be use for such cases).

Reply:
Ok, fixed, see below.